kratos/modeler/operation/find_contacts_in_skin_model_part.cpp

```diff
@@ -0,0 +1,250 @@
+    array_1d<std::size_t, 3> max_position(3,0);
+    std::vector<int> face_neigh_colors;
+    face_neigh_colors.reserve(6);
+    std::vector<int>  all_neigh_colors;
+    all_neigh_colors.reserve(26);
+    const CartesianMeshColors& r_colors = GetMeshColors();
+    r_colors.CalculateOuterMinMaxNodePositions(rCondition.GetGeometry(), min_position, max_position);
+    for(std::size_t k = min_position[2] ; k < max_position[2] ; k++){
+        for(std::size_t j = min_position[1] ; j < max_position[1] ; j++){
+            for(std::size_t i = min_position[0] ; i < max_position[0] ; i++){
+                Point cell_min_point = r_colors.GetPoint(i,j,k);
+                Point cell_max_point = r_colors.GetPoint(i+1,j+1,k+1);
+                const int current_cell_color = std::lround(r_colors.GetElementalColor(i,j,k));
+                if (current_cell_color == CellColor &&
+                    rCondition.GetGeometry().HasIntersection(cell_min_point,cell_max_point))
+                {
+                    //This cell intersects with condition and has correct color
+                    const array_1d<std::size_t, 3> indexes{i,j,k};
+                    std::vector<array_1d<std::size_t, 3>> face_neigh_indexes;
+                    GetCellNeihgbourColorsConnectedByFace(indexes, r_colors, face_neigh_colors, face_neigh_indexes);
+
+                    // We check if there is any neighbour that satisfies:
+                    // is connected by face, projection is within certain angle, distance
+                    // to condition is minimum.
+                    GetClosestContactColorFromNeighbours(
+                        rContactContainer.mColorVector,
+                        face_neigh_colors,
+                        face_neigh_indexes,
+                        condition_center,
+                        condition_normal,
+                        r_colors,
+                        pairing_status,
+                        min_distance,
+                        temporary_color);
+
+                    // if not contact by face has been found we look for contacts by corner.
+                    if (pairing_status < 2) {
+                        const array_1d<std::size_t,3> indexes{i,j,k};
+                        GetCellNeihgbourColors(indexes,all_neigh_colors, r_colors);
+                        for (const auto& contact_color : rContactContainer.mColorVector) {
+                            if (std::find(all_neigh_colors.begin(), all_neigh_colors.end(),contact_color) != all_neigh_colors.end()) {
+                                pairing_status = 1;
+                                temporary_color = contact_color;
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    if (pairing_status > 0) {
+        //We did find a temporary candidate , so we use this one
+        KRATOS_CRITICAL_SECTION
```

Comment:
I assume this is here because this code is called within a parallel section? (Or will be in the future)

Reply:
I pointed where